### ISSUE TITLE
New item

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8336]
As a court, so that motions can be granted or denied expeditiously, I need the ability to "stamp" grant or deny motions filed with the court.

When motions are filed with the court, a judge must grant or deny the motion. In prior systems, commonly filed motions were "stamp" granted or denied, meaning that instead of writing out additional documentation (for example, an order explicitly granting or denying the motion) a judge or a member of their chambers could simply stamp "Granted" or "Denied" on the motion itself and make an accompanying short docket entry explaining the terms of said disposition. This made processing motions, especially common ones, very fast with a minimum of extra work needed to create additional documentation.

In DAWSON, we would like a mechanism to perform this same function for the same reasons; a way for a judge or their chambers to "stamp" grant or deny motions with a minimum of effort. 

The workflow should be able to be triggered for a judge, chambers, or ADC user in the context of a motion. The user creating the stamp disposition should be able to specify the disposition by selecting details appropriate to the disposition and potentially adding text of their own. The combination of these details and any additional text will be used to create an entry on the docket record. The entry on the docket will be accompanied by a copy of the cover sheet to the original motion that includes a new stamp, along with the judge's signature and be coded as event type "O" (general order).

## Pre-Conditions
ustaxcourt/ef-cms#8339

## Acceptance Criteria
* For a motion, a judge/chambers/ADC user can choose to enter a stamp disposition for the motion
* This functionality can be invoked from either the Document View or the Attachment view (when viewing a message to which the Motion is attached)
* When entering a stamp disposition:
  - user **must** choose one (and only one) of the following selections:
      - Granted
      - Denied
          - if "Denied" is selected, then the user _may_ also choose one or both of the following options:
              - Without prejudice
              - As moot
 - The user _may_ also select options to include other text as applicable:
  - "This case is stricken from the trial session"
  - EITHER:
    - "This case is restored to the general docket"
    - "Jurisdiction is retained by the undersigned"
  - EITHER:
     - "The parties shall file a status report by [DATE]" - ideally w/ a date picker to select the date and validation to prevent selecting dates prior to [TODAY]
     - "The parties shall file a status report or proposed stipulated decision by [DATE]" - ideally w/ a date picker to select the date and validation to prevent selecting dates prior to [TODAY]
  - user can also add free text up to 60 characters in length in a field labeled "Custom Order Text (optional)"
  - user has option to Save or Cancel
  - ideally the user can preview the stamp as they make selections in the UI
* Upon saving the stamp disposition, a document is presented (henceforth referred to as a Stamp Order). The document:
  - is a copy of the **cover sheet only** of the original motion
  - has a stamp on the cover sheet containing:
       - the first line reading ""It is ORDERED as follows:"
       - the input supplied by the user when entering the stamp disposition printed on it (predefined choice selection(s) and additional free text segregated line by line per option) - see UX mockups/comments for visual
       - the signature of the appropriate judge
* The user has the opportunity to save, or delete the Stamp Order as they would any other Draft
   - Deleting the Stamp Order asks the user to confirm that they want to delete the Stamp Order draft, after which it is deleted
* After saving, the Stamp Order appears as a Draft attached to the case.
* A Docket user can select the Draft and choose to Add Docket Entry.
* When the Docket user clicks Add Docket Entry:
  - The Description field is populated with the title of the motion (e.g., "Motion for Continuance") followed by the text of the Stamp options selected during creation, concatenated
  - Docket clerk can edit the Description field as usual
* When the Stamp Order is added to the docket:
  - The docket entry has event code "O"
  - The docket entry displays the text of the applied stamp along with the type of Motion being disposed of in the "Filings and Proceedings" column on the docket record
  - The docket entry is hyperlinked to the Stamp Order
  - The docket entry obeys all usual visibility rules
  - The docket entry is served to all parties in the usual manner (including presenting the Stamp Order for printing in the case where parties receive paper service) and receives a service stamp at the bottom with the service date as usual

## Notes
* What format should the stamp disposition take on the Stamp Order? - UX

Example of this functionality from Blackstone (1111-18, Dkt. ustaxcourt/ef-cms-flexion#9):

![image.png](https://images.zenhubusercontent.com/60e7a3b441280d0c68e72525/af735045-4891-419b-ac15-2616609ae5f8)

## Tasks
- [x] Add 'Apply Stamp' button to document view - Rachel (https://github.com/flexion/ef-cms/pull/9527)
- [x] Add 'Apply Stamp' button to messages view - Rachel (https://github.com/flexion/ef-cms/pull/9527)
- [x] Route to Apply Stamp form when button is clicked - Rosie + Rachel (https://github.com/flexion/ef-cms/pull/9529)
- [x] Apply Stamp form - Rosie + Rachel
  - [x] Styling of stamp - Rosie (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Create ApplyStamp component to contain the form and PDF preview - Rosie (https://github.com/flexion/ef-cms/pull/9529) 
  - [x] Make form controls update the text in the stamp - Rosie (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Dynamically render the character count for custom order text - Rosie (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Disable checkboxes unless "Denied" is selected - Rosie (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Clear selected checkboxes under "Denied" if the user selects Granted instead - Rosie (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Statically place stamp underneath document title, happy path 1 page coversheet - Rosie (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Revert static placement of stamp and allow user to place stamp on hover - Rosie + Rachel (https://github.com/flexion/ef-cms/pull/9530)
  - [x] Reset optional radio buttons with a clear button - Rosie (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Match wireframe styling, which is based on Document View - Rosie (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Ensure DatePicker values are dynamically updated and cleared if switching radio options - Rosie + Mrinalini + Rachel (https://github.com/flexion/ef-cms/pull/9529)
  - [x] Only display the coversheet in the PDF preview on Apply Stamp page - Rosie + Rachel (https://github.com/flexion/ef-cms/pull/9530)
  - [x] Add validation to prevent selecting dates prior to [TODAY] in both DatePickers - Rosie + Rachel (https://github.com/flexion/ef-cms/pull/9537)
  - [x] Add validation on "Save Stamp Order" for the form - Rosie + Rachel + Mrinalini (https://github.com/flexion/ef-cms/pull/9537)
  - [x] Disabled "Save Stamp Order" button until Granted/Denied is selected - Rosie + Rachel (https://github.com/flexion/ef-cms/pull/9531)
- [x] Apply stamp using stamp + form data on to motion coversheet -- ALL other tasks dependent on this one - Mrinalini, Rosie https://github.com/flexion/ef-cms/pull/9551
  - [x] update stamp entity to have all properties form data - Mrinalini, Rosie, James https://github.com/flexion/ef-cms/pull/9543
- [x] make sure clear all fields clears ALL optional form fields - Rosie - https://github.com/flexion/ef-cms/pull/9551
- [x] generate draft stamp order with motion coversheet as pdf contents - Mrinalini, Rosie https://github.com/flexion/ef-cms/pull/9543
- [x] delete functionality for draft stamp order
  - [x] delete UI + sequence
  - [x] delete backend code
- [x] add docket entry functionality for serving stamp order - Mrinalini, Rosie https://github.com/flexion/ef-cms/pull/9543
- [x] verify on click and save of apply stamp for motion that already has one generates ANOTHER stamp order (as many as u want) https://github.com/flexion/ef-cms/pull/9543
- [x] remove edit and remove stamp buttons on draft - Mrinalini, Rosie https://github.com/flexion/ef-cms/pull/9551
- [x] Verify "A Docket user can select the Draft and choose to Add Docket Entry" - Rosie
- [x] The Description field is populated with the title of the motion (e.g., "Motion for Continuance") followed by the text of the Stamp options selected during creation, concatenated https://github.com/flexion/ef-cms/pull/9552
- [x] Add the docket entry and populate all fields as per the AC https://github.com/flexion/ef-cms/pull/9552
- [x] Message flow - starting on the messages page, then apply the stamp - return to messages instead of navigating to drafts. - Rosie https://github.com/flexion/ef-cms/pull/9557
- [x] Add Automated testing: `generateCoverSheetData`, `generateStampedCoversheetInteractor` - Rosie https://github.com/flexion/ef-cms/pull/9557
- [x] Clear optional button should empty out the customText box - James https://github.com/flexion/ef-cms/pull/9556
- [x] Form clearing: if you set the date, then clear, and reset the same date field it doesn't work - Cody https://github.com/flexion/ef-cms/pull/9556
- [x] If you hit the back button from messages and there are multiple attachments, you should go back to the motion that you came from (should be the one that has Apply Stamp) not the latest attachment. This is probably a really small edge case. - James, Fritz, Mrin ustaxcourt/ef-cms-flexion#9561
DOD
- [x] PDF visual test with the generated stamped coversheet
Integration testing
- [x]  Verify the behavior of the back button for Messages and Document Preview (see https://github.com/flexion/ef-cms/pull/9552#discussion_r931228665)
- [x]  Consider combining endpoint calls in completeMotionStampingAction to increase performance
- [x]  Consider if we need a Stamp entity just for validation. Since it's being passed through as part of the docket entry, maybe the entity is not needed.
- [x]  Revisit height adjustment of docket-entry-preview-text to not be hardcoded - talk to UX
- [x]  Does the back button on the stamp form need to have specific words for which page you came from?
- [x]  If there's logic in ApplyStamp.jsx, move it to a helper
- [x]  Test rawDocketEntry.stampData defaults to {} - mrinalini
- [x]  Validate Stamp before putting it on the DocketEntry - mrinalini
- [x]  Find out from users if two served stamps on the stamp order is confusing (might just wait for user feedback) --mrinalini ustaxcourt/ef-cms-flexion#9558

- [ ] PR to remove feature flag - (https://github.com/ustaxcourt/ef-cms/pull/2866)

## Story Definition of Ready (Created on 9/26/21)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ]  Be immediately actionable for the development team
- [ ]  Be focused on users
- [ ]  Include a narrative. “As a ____, so that I can ______, I need _____.”
- [ ]  Include user-centric acceptance criteria that expresses the users' needs
- [ ]  Include user-centric testing scenarios
- [ ]  Include preconditions (story dependencies)
- [ ]  Not include design or technical solutions
- [ ]  Include non-standard testing, deployment, documentation, browser, mobile or security requirements, as well as any requirements for 3rd-party integrations

Process: Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 10-6-21)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [x] Automated test scripts have been written
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Acceptance criteria for the story has been met
 - [x] If there are special instructions in order to deploy into the next environment, add them as a comment in the story
 - [x] Deployed to any Flexion environment for Engineering Review
 - [x] Deployed to the Court's migration environment

**Comment by swongCO** [comment]
Figma link: https://www.figma.com/file/81RYWvF84vxJBBnfHBWR1U/USTC-%7C-Drafts-%7C-Stamp-Disposition-9479?node-id=95%3A608

![Apply stamp - sidebar preview default state.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/790748c8-c07e-4f40-adde-fbfffcc59a49)
![Apply stamp - sidebar preview.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/3e7e3c3f-630d-44c9-a73f-bd5ade1861d9)

1. Follow existing design patterns, interactions, flow and layout for title and all components except where noted
2. Use Document View as guide for layout of sidebar and document window
3. Sidebar = 5 columns; 20 px right padding
4. Document window = 7 columns; 20 px left padding
5. Back to link should return user to previous page (currently Document View or Message/Case Messages in the docket nav)
6. Pagination component - align left with document window
7. In the optional selections, radio btns can be deselected if clicked
8. Note the date input design in the body text; functionality/validation is the same as standard date picker
9. Apply Stamp btn is disabled until the first input is entered
10. Save Stamp Order btn is disabled until stamp is applied
11. Remove Stamp link is contingent on ustaxcourt/ef-cms-flexion#14
12. Create the stamp based on the user's inputs - width and height will be responsive to the inputs
13. **NOT REQUIRED!** Display the stamp In the Stamp Preview window - determine if it can be a live preview or if it requires a trigger to render
14. Apply Stamp - determine if the stamp can automatically be placed into the document view in the mocked up position
15. Determine the flow if there is more than one page in the document
16. Size of the stamp on the document will be determined by the margins and visual aesthetics on screen and print 

Stamp examples:
![Stamp - granted.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/069c5a18-4bd9-4157-9c06-ced6532100fe)
![Stamp - denied.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/b6539bce-fa26-47b8-83a0-ebebdf65793c)

**Comment by swongCO** [comment]
![Docket record - apply stamp.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/c299cc4e-f290-4c0c-b52a-38286b3fb658)
![Message view - apply stamp.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/5ff59ad9-1014-46a3-965d-22b4dff7d88c)
1. Add apply stamp link in the blue banner atop the document view window
2. Use the stamp icon from fontAwesome
3. Clicking the link will display the Apply Stamp page

**Comment by swongCO** [comment]
![Drafts view.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/ca0b9822-f836-45dd-b281-b0cef23dfccd)
![Drafts view modal.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/23416c14-c244-4749-8a4f-3c8275af8468)
1. In the Drafts page, the stamp order can be edited or deleted by the creator
2. If Edit is selected, the user will be taken to the Apply Stamp page where their inputs are pre-populated
3. If Delete is selected, display the modal

**Comment by ttlenard** [comment]
## Test Cases

Notes
- _Test cases 1-10 test Stamp Order functionality via the Message workflow._
- _Test cases 13-15 test Stamp Order functionality via the Document Preview workflow._
- _Test cases 16-17 test Stamp Order Search and public terminal user functionality_

#### 1) Motion is filed by a party; Docket clerk QC's motion; sends message with the motion attached to Judge; Judge/CA/ADC user reviews message; Apply Stamp button is available on the document preview from the message thread.

This test case will take some set up
- Log in to DAWSON as a Party on a case (could be petitioner, Private Practitioner, or IRS Practitioner). 
- File a document in a case
- For the document type, select any type of Motion, and then click continue.
- Upload the document, fill in the required fields, and Submit your filing
- Next, log in as a docket clerk.
- Navigate to the case you just submitted the motion in.
- Click on the document link to QC the Motion.
- Click "Complete QC", and then click "Complete and Send Message"
- Select Chambers/Buch's Chambers/Judge Buch, add in a test message, and click Send.

Now the test is ready
- Log in as Judge Buch
- Click on the new message from Docket re: the motion you submitted (it should be the first message in Recent Messages on your dashboard).

Expected Results
- [x] **NEW FUNCTIONALITY** - The apply stamp button is available on the document preview from the Message/Attachments page.

-Repeat this test by having docket send a message with a motion attached to an ADC user, and CA user. Next log in as those users to ensure that these users also have access to the "Apply Stamp" button.



#### 2) Judge/CA/ADC User applies stamp on a motion from a message attachment; Judge/CA/ADC User is presented with the Apply Stamp Page.

- Continuing from the previous test
- Click on the "Apply Stamp" button

Expected Results
- [x] **NEW FUNCTIONALITY** - The new Apply Stamp page is available to the user

- Repeat this test with ADC and CA user



#### 3) Judge/CA/ADC User begins to fill out stamp information on the apply stamp page, Selects the "Back" link; information on the apply stamp page is essentially cancelled, and not saved; User is routed back to their message view.

- Continuing from the previous test
- Begin to select some of the fields/add some additional order text
- Click on the link "Back" link on the top left of the "Apply Stamp" screen

Expected Results
- [x] **NEW FUNCTIONALITY** - The user is directed back to the Message attachments page, where they can see the "Apply Stamp" button on the document preview.

- Repeat this test with the ADC and CA user.



#### 4) Judge/CA/ADC User fills out stamp information on the apply stamp page, rules for each field follow AC

- Continuing from the previous test
- Click on the "Apply Stamp" button
- On the "Apply Stamp" Page, Add in some Stamp details. 
- Check the functionality of each of the radio buttons, additional order text field, and the Stamp Preview

Expected Results:
- [x] **NEW FUNCTIONALITY** - Save Stamp Order button at the top of the page does not "activate" unless Granted or Denied is selected.
- [x] **NEW FUNCTIONALITY** - The radio button for "This case is stricken from the trial session" is optional, and can be selected/unselected'
- [x] **NEW FUNCTIONALITY** - You can select/deselect any of the radio buttons underneath the Select any that apply section.
- [x] **NEW FUNCTIONALITY** - You can enter a date for today or in the future by either typing in a date, or selecting a date from a date picker
- [x] **NEW FUNCTIONALITY** - You can NOT enter a date in the past by either typing in a date, or selecting a date from a date picker (or if you can enter a date in the past, a validation error appears)
- [x] **NEW FUNCTIONALITY** - Custom order text box allows for 60 characters only
- [x] **NEW FUNCTIONALITY** - All information that is entered/selected in the Stamp input area is dynamically able to be previewed in the "Stamp Preview" box, and includes the appropriate Judge's Signature.

- Repeat this test with the ADC and CA user



#### 5) Judge/CA/ADC User clicks the "Save Stamp Order"; Order is saved to the drafts folder of the case 

- Continuing from the previous test
- Add in the Stamp Order Details 
- Click the button to "Save Stamp Order"


- [x] **NEW FUNCTIONALITY** - The signed stamp order is available in the Drafts folder (you should see the number icon indicating a new draft)
- [x] **NEW FUNCTIONALITY** - The title/description of the draft Stamp Order should include all data selected/entered from the Stamp Order area


- Repeat this test with the ADC and CA user


#### 6) Judge/CA/ADC user clicks the "Delete" button on the Drafts Screen, Modal pops up asking the user to confirm the deletion of the Stamp Order, User selects "No, Take Me Back"

- Continuing from the previous test, View the stamp order from the drafts folder
- Click the Delete Button

Expected Results
- [x] **NEW FUNCTIONALITY** - A Modal will pop up asking you to confirm "Are you sure you want to delete this document? Once deleted, it cant be restored."

- Select "No, Take me back"

Expected Results
- [x] **NEW FUNCTIONALITY** - The User is navigated back to the drafts page where they can preview the document, have the option to delete, or View PDF.

- Repeat this test with the ADC and CA user


#### 7) Judge/CA/ADC user clicks the "Delete" button on the Drafts Screen, Modal pops up asking the user to confirm the deletion of the Stamp Order, User selects "Yes, Delete"

- Continuing from the previous test
- Click the Delete Button

Expected Results
- [x] **NEW FUNCTIONALITY** - A Modal will pop up asking you to confirm that you "Are you sure you want to delete this document? Once deleted, it can't be restored.."

- Select "Yes, Delete"

Expected Results:
- [x] **NEW FUNCTIONALITY** - The Stamp Order is deleted
- [x] **NEW FUNCTIONALITY** - User sees the "Apply Stamp" button on the original Motion from the message attachment.

- Repeat this test with the ADC and CA user


#### 8) Judge/CA/ADC user Creates a Stamp Order on a Motion, Applies Stamp, Saves Stamp Order; User can attach the draft stamp order to a Message Thread

- Continuing from the previous test
- Click the "Apply Stamp" button on the Original Motion from the Message Attachment
- Add in the Stamp Order details
- Click Apply Stamp
- Click Save Stamp Order
- Navigate back to the message, click reply


Expected Results:
- [x] The Message Modal has the draft Stamp Order attached



#### 9) Judge/CA/ADC User Sends Message to Docket; Docket Adds Stamp Order to the Docket Record

- Continuing from the previous test
- Send the message with the draft Stamp Order to Docket
- Log in as a Docket Clerk
- Navigate to the message
- Click "Add Docket Entry"

Expected Results
- [x] **NEW FUNCTIONALITY** - Description field is populated with the title of the motion followed by the text of the Stamp options selected
- [x] **NEW FUNCTIONALITY** - Description field can be edited by the docket clerk


#### 10) Docket Clerk Adds the Stamp Order to the Docket Record

- Continuing from the previous test
- Docket Clerk clicks "Add Docket Entry"

Expected Results
- [x]  **NEW FUNCTIONALITY** - The docket entry has event code "O"
- [x]  **NEW FUNCTIONALITY** - The Filings and Proceedings Column displays the type of motion along with the text of the entire Stamp Order
- [x]  **NEW FUNCTIONALITY** - The docket entry is hyperlinked
- [x]  **NEW FUNCTIONALITY** - The docket entry is served to all parties in the usual manner, ex: if the parties have paper service, the print paper service page is presented.



#### 11) Motion is filed by a party; Docket clerk QC's motion; Judge/CA/ADC user reviews docket record, Clicks on Document View of the Motion; Apply Stamp button is available on the document preview screen.

This test case will take some set up
- Log in to DAWSON as a Party on a case (could be petitioner, Private Practitioner, or IRS Practitioner). 
- File a document on a case
- For the document type, select any type of Motion, and then click continue.
- Upload the document, fill in the required fields, and submit the filing
- Next, log in as a docket clerk.
- Navigate to the case you just submitted the motion in.
- Click on the document link to QC the Motion.
- Click "Complete QC"

Now the test is ready
- Log in as Judge Buch
- Navigate to the QC'd Motion from above.
- Click on the document link to get to the Document View

Expected Results
- [x] **NEW FUNCTIONALITY** - The apply stamp button is available on the Document View page



#### 12) Judge/CA/ADC User clicks apply stamp on a motion from the document view; Judge/CA/ADC User is presented with the Apply Stamp Page.

- Continuing from the previous test
- Click on the "Apply Stamp" button

Expected Results
- [x] **NEW FUNCTIONALITY** - The new Apply Stamp page is available to the user

- Repeat this test with ADC and CA user


#### 13) Judge/CA/ADC User begins to fill out stamp information on the apply stamp page, Selects the "Back" link; information on the apply stamp page is essentially cancelled, and not saved; User is routed back to the Document view page.

- Continuing from the previous test
- Begin to select some of the fields/add some additional order text
- Click on the "Back" link on the top left of the "Apply Stamp" screen

Expected Results
- [x] **NEW FUNCTIONALITY** - The user is directed back to the Document View page, where they can see the "Apply Stamp" button on the document preview.

- Repeat this test with the ADC and CA user.


#### 14) Judge/CA/ADC User fills out stamp information on the apply stamp page, rules for each field follow AC

- Continuing from the previous test
- Click on the "Apply Stamp" button
- On the "Apply Stamp" Page, Add in some Stamp details. 
- Check the functionality of each of the radio buttons, additional order text field, and the Stamp Preview

Expected Results:
- [x] **NEW FUNCTIONALITY** - Apply Stamp button at the bottom of the page does not "activate" unless Granted or Denied is selected.
- [x] **NEW FUNCTIONALITY** - The radio button for "This case is stricken from the trial session" is optional, and can be selected/unselected'
- [x] **NEW FUNCTIONALITY** - You can select/deselect any of the radio buttons underneath Select Any that Apply section.
- [x] **NEW FUNCTIONALITY** - You can enter a date for today or in the future by either typing in a date, or selecting a date from a date picker
- [x] **NEW FUNCTIONALITY** - You can NOT enter a date in the past by either typing in a date, or selecting a date from a date picker (or if you can enter a date in the past, a validation error appears)
- [x] **NEW FUNCTIONALITY** - Custom order text box allows for 60 characters only
- [x] **NEW FUNCTIONALITY** -  All information that is entered/selected in the Stamp input area is dynamically able to be previewed in the "Stamp Preview" box, and includes the appropriate Judge's Signature.

- Repeat this test with the ADC and CA user



#### 15) Judge/CA/ADC User clicks the "Save Stamp Order"; Order is saved to the drafts folder of the case. 

- Continuing from the previous test
- Add in the Stamp Order details
- Click the button to "Save Stamp Order"

Expected Results:
- [x] **NEW FUNCTIONALITY** - The signed stamp order is available in the Drafts folder (you should see the number icon indicating a new draft)
- [x] **NEW FUNCTIONALITY** - The title/description of the draft stamp order should include all inputs selected from the Stamp Order Details

- Repeat this test with the ADC and CA user


#### 16) Regression test searching for these Stamp Orders using the Order Search page
- Perform an Order Search as a logged in Court user

Expected Results
- [x] Stamp Order (sealed and unsealed) returned in results when searching by Judge
- [x] Stamp Order (sealed and unsealed) returned when searching by date
- [x] Stamp Order (sealed and unsealed) returned when searching Case title/petitioner name
- [x] Stamp Order (sealed and unsealed) returned when searching by docket number
- [x] Stamp Order (sealed and unsealed) returned when searching by keyword or phrase (example: Denied or Granted)

- Perform an Order Search as a party to a case (IRS, Private Practitioner)

Expected Results
- [x] Stamp Order (sealed to the public and unsealed) returned in results when searching by Judge
- [x] Stamp Order (sealed to the public and unsealed) returned when searching by date
- [x] Stamp Order (sealed to the public and unsealed) returned when searching Case title/petitioner name
- [x] Stamp Order (sealed to the public and unsealed) returned when searching by docket number
- [x] Stamp Order (sealed to the public and unsealed) returned when searching by keyword or phrase (example: Denied or Granted)


- Perform an Order Search as a non-logged in user

Expected Results
- [ ] Stamp Order (sealed) is not returned in results when searching by Judge
- [ ] Stamp Order (unsealed) returned when searching by date
- [ ] Stamp Order (unsealed) returned when searching Case title/petitioner name
- [ ] Stamp Order (unsealed) returned when searching by docket number
- [ ] Stamp Order (unsealed) returned when searching by keyword or phrase (example: Denied or Granted)

- Perform an Order Search as a user on the Public Terminal 

- [x] Stamp Order (sealed) is not returned in results when searching by Judge
- [x] Stamp Order (unsealed) returned when searching by date
- [x] Stamp Order (unsealed) returned when searching Case title/petitioner name
- [x] Stamp Order (unsealed) returned when searching by docket number
- [x] Stamp Order (unsealed) returned when searching by keyword or phrase (example: Denied or Granted)


#### 17) Regression test - Public Terminal User can open and view Stamp Orders that on the Docket Record
- Log in as the public terminal user
- Navigate to a case that has a stamp order (Sealed and Unsealed)

Expected Results
- [x] Can open and view Stamp Orders that are not sealed
- [x] Sealed stamp order is visible on the docket record, but there isn't a link top open it

**Comment by cholly75** [comment]
@swongCO @liadmccabe - one additional piece of feedback from the Court: for the stamp itself, they would like the first line to read "The Court orders as follows:" by itself, and move "This motion is" to the second line along with the Granted/Denied text.

**Comment by cholly75** [comment]
@swongCO @liadmccabe Remember the comment I made 5 hours ago saying the first line should read "The Court orders as follows:"? Man, those were good times. Heady days, indeed.

Anyhoo, strike that. It should read, instead, "It is ORDERED as follows:"

Thanks.

**Comment by rosiehsmith** [comment]
@cholly75 per our discussion - the `edit` functionality and all AC related to it will be split out into another story/vertical slice. Current story slice includes applying the stamp, saving then generating a draft Stamp Order, ability to delete a draft Stamp Order, ability to add docket entry and serve Stamp Order

**Comment by cholly75** [comment]
@rosiehsmith @ttlenard - removed mentions of Edit functionality w/r/t the Stamp Order in the story AC.

**Comment by swongCO** [comment]
I haven't gone through the test cases yet but did preliminary testing with an existing case/motion.

1. When I open up the date picker calendar and click on the arrows to go to the next month/year, it exits the calendar back to the radio btn in focus
2. When I entered a date with the wrong format, it didn't trigger an error state (screenshot)
![Screen Shot 2022-08-08 at 3.35.56 PM.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/f5a2750c-cbe3-4db3-a3fb-98d9d6a44629)
3. When I entered a date with the slightly different wrong format, it triggered an incorrect error message - should be that the format is wrong, not that it's a prior date (screenshot)
![Screen Shot 2022-08-08 at 3.37.41 PM.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/6ab67ad3-7a80-46ef-a3b1-b395d9b19de9)
4. The stamp is overlapping content on the cover page so the positioning of the stamp will need to be adjusted (screenshot)
![Screen Shot 2022-08-08 at 3.41.37 PM.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/4761712c-475b-485c-ae06-45933fbe10f2)
5. Change the following font sizes/weight to create visual hierarchy and to emphasize action items:
  a. "This motion is...", including the two options for Denied, should be 24px, instead of 16px which is the current body font size
  b. DENIED/GRANTED should be 28px (keep them bolded)
  c. The options with the dates should be bolded if selected
  d. The stamp should look something like this:
![Stamp - denied.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/d22f9ff6-ffa1-4e5e-b6ac-da5236d9d89a)
Let me know if there are any questions!

**Comment by jamesobrooks** [comment]
@swongCO for item 5 in your latest comment, feedback we got from the court was to _not_ emphasize anything. @mmarcotte, and @ttlenard will have more context. I can check out items 1-3.

**Comment by jamesobrooks** [comment]
Item 2:
Note that the stamp has the correct date despite the input having dashes rather than slashes.
![Screen Shot 2022-08-08 at 6 29 25 PM](https://user-images.githubusercontent.com/12275865/183531710-f3dde43b-c0d8-4555-8e20-6ab6022d23a9.png)

Item 3:
We can check on how to make this work potentially, but we are using the date picker from USWDS and it is converting that to a date in the first century. See this from my logs locally:
```
JOB:: 0022-09-21T04:56:02.000Z
```

**Comment by jamesobrooks** [comment]
Actually, it looks like potentially the input area is ours, but then we need to figure out a way to handle short years that makes sense. For example, I can add this code:
```
if (year.length < 4) {
  year = '';
}
```

This does trigger the `Enter a valid date` validation error. But we already have code that handles entering the date in various formats which might get a little weird. For example, we do accept dates like `2022-12-01` in addition to `12-01-2022`. We do this by checking the length of the first token after splitting, so if > 2, we assume it is the year.

Either way we go, if the date is 6 numbers with 2 identical separators, it probably doesn't matter that much if the user entered `12-01-22` or `22-12-01` as it will be rejected. The other option I can see right now would be assuming the year should start with '20' and prepending that to a 2-digit year, but I'm not a big fan of that.

**Comment by cholly75** [comment]
@swongCO @jamesobrooks - re: Sui's feedback: items 1/2/3 should be addressed in some manner. Item 4 - that's an old Blackstone motion in the screenshot example, can we see how it looks with a DAWSON-generated motion cover sheet (it looks correct in James' screenshot to me)? Item 5 - James is correct, the request is to remove the styling altogether.

Thanks,
Chris

**Comment by cholly75** [comment]
Item: "Created By" text in draft view for stamp order is displaying the creator/filer of the original motion rather than the creator of the Stamp order:

![image.png](https://images.zenhubusercontent.com/60e7a3b441280d0c68e72525/920e3c0a-ddfc-4368-a7cf-a8f3b6c27529)

**Comment by rosiehsmith** [comment]
@mmarcotte @jimlerza since this story includes the [feature flag addition](https://github.com/ustaxcourt/ef-cms/pull/2437), the manual deploy step required is:

Run script `./enable-stamp-disposition.sh`  to setup the `stamp-disposition-enabled` feature flag (default value being true).
